x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java

```diff
@@ -0,0 +1,256 @@
+
+        assertThat(blockingFuture.isDone(), is(false));
+        for (PlainActionFuture<Integer> future : supersededFutures) {
+            assertThat(future.isDone(), is(false));
+        }
+        assertThat(responseFuture.isDone(), is(false));
+
+        // release the initial blocked request.
+        await(barrier);
+
+        assertThat(blockingFuture.actionGet(), equalTo(0));
+        supersededFutures.forEach(future -> assertThat(future.actionGet(), equalTo(response)));
+        assertThat(responseFuture.actionGet(), equalTo(response));
+
+        assertThat(responses, empty());
+        assertBusy(() -> assertThat(cache.jobQueueSize(), equalTo(0)));
+        assertThat(cache.jobQueueCount(), equalTo(0));
+    }
+
+    public void testMultipleRequestsCancelled() throws Exception {
+        Queue<Function<Runnable, Integer>> responses = new ArrayDeque<>();
+        CapacityResponseCache<Integer> cache = new CapacityResponseCache<>(
+            runOnThread,
+            ensureNotCancelled -> responses.remove().apply(ensureNotCancelled)
+        );
+
+        CyclicBarrier barrier = new CyclicBarrier(2);
+        responses.add(ensureNotCancelled -> {
+            await(barrier);
+            await(barrier);
+            return 0;
+        });
+
+        PlainActionFuture<Integer> blockingFuture = new PlainActionFuture<>();
+        cache.get(() -> false, blockingFuture);
+        await(barrier);
+
+        AtomicBoolean cancelled = new AtomicBoolean();
+        int count = between(1, 10);
+        List<PlainActionFuture<Integer>> supersededFutures = new ArrayList<>(count + 1);
+        for (int i = 0; i < count; ++i) {
+            PlainActionFuture<Integer> supersededFuture = new PlainActionFuture<>();
+            supersededFutures.add(supersededFuture);
+            cache.get(cancelled::get, supersededFuture);
+        }
+
+        CyclicBarrier responseBarrier = new CyclicBarrier(2);
+        responses.add(ensureNotCancelled -> {
+            await(responseBarrier);
+            await(responseBarrier);
+            ensureNotCancelled.run();
+            fail();
+            return 0;
+        });
+        PlainActionFuture<Integer> responseFuture = new PlainActionFuture<>();
+        cache.get(cancelled::get, responseFuture);
+
+        assertThat(blockingFuture.isDone(), is(false));
+        for (PlainActionFuture<Integer> future : supersededFutures) {
+            assertThat(future.isDone(), is(false));
+        }
+        assertThat(responseFuture.isDone(), is(false));
+
+        // release the first blocking thread.
+        await(barrier);
+        assertThat(blockingFuture.actionGet(), equalTo(0));
+
+        await(responseBarrier);
+        cancelled.set(true);
+        await(responseBarrier);
+        supersededFutures.forEach(future -> expectThrows(TaskCancelledException.class, future::actionGet));
+        expectThrows(TaskCancelledException.class, responseFuture::actionGet);
```

Comment:
Could we also have a test that verifies that if a strict subset of the jobs are cancelled then execution still proceeds?

Reply:
I added this in the prior commit/test by [randomly letting the superseded tasks cancel](https://github.com/elastic/elasticsearch/pull/80499/commits/ed38bcb31f681d0297dec34f7b0bf9cd36a3c476#diff-84ea97ed99c61e1fcb20519981e9ec441f105817d74211ae7b6db06fe3387d61R127), I think that is good enough?

Reply:
Yep looks good :)